jdbc/src/java/org/apache/hive/jdbc/HiveConnection.java

```diff
@@ -896,39 +946,103 @@ private void openSession() throws SQLException {
+    for (int i=0; i<numRetry; i++) {
+      try {
+        openSession(openReq);
+      } catch (TException e) {
+        if (isSamlRedirect(e)) {
+          boolean success = doBrowserSSO();
+          if (!success) {
+            String msg = browserClient.getServerResponse() == null
+                || browserClient.getServerResponse().getMsg() == null ? ""
+                : browserClient.getServerResponse().getMsg();
+            throw new SQLException(
+                "Could not establish connection to " + jdbcUriString + ": "
+                    + msg, " 08S01", e);
+          }
+        } else {
+          throw new SQLException(
+              "Could not establish connection to " + jdbcUriString + ": " + e
+                  .getMessage(), " 08S01", e);
+        }
+      }
+    }
+    isClosed = false;
+  }
+
+  private boolean doBrowserSSO() throws SQLException {
     try {
-      TOpenSessionResp openResp = client.OpenSession(openReq);
+      Preconditions.checkNotNull(browserClient);
+      try (IJdbcBrowserClient bc = browserClient) {
+        browserClient.doBrowserSSO();
+        HiveJdbcBrowserServerResponse response = browserClient.getServerResponse();
+        if (response != null) {
+          return response.isSuccessful();
+        }
+        return false;
+      }
+    } catch (Exception ex) {
+      throw new SQLException("Browser based SSO failed: " + ex.getMessage(),
+          " 08S01",
+          ex);
+    }
+  }
 
-      // Populate a given configuration from HS2 server HiveConf, only if that configuration
-      // is not already present in Connection parameter HiveConf i.e., client side configuration
-      // takes precedence over the server side configuration.
-      Map<String, String> serverHiveConf = openResp.getConfiguration();
+  @VisibleForTesting
+  public IJdbcBrowserClient getBrowserClient() {
+    return browserClient;
+  }
 
-      updateServerHiveConf(serverHiveConf, connParams);
+  private void openSession(TOpenSessionReq openReq) throws TException, SQLException {
+    TOpenSessionResp openResp = client.OpenSession(openReq);
 
-      // validate connection
-      Utils.verifySuccess(openResp.getStatus());
-      if (!supportedProtocols.contains(openResp.getServerProtocolVersion())) {
-        throw new TException("Unsupported Hive2 protocol");
-      }
-      protocol = openResp.getServerProtocolVersion();
-      sessHandle = openResp.getSessionHandle();
-
-      final String serverFetchSizeString =
-          openResp.getConfiguration().get(ConfVars.HIVE_SERVER2_THRIFT_RESULTSET_DEFAULT_FETCH_SIZE.varname);
-      if (serverFetchSizeString == null) {
-        throw new IllegalStateException("Server returned a null default fetch size. Check that "
-            + ConfVars.HIVE_SERVER2_THRIFT_RESULTSET_DEFAULT_FETCH_SIZE.varname + " is configured correctly.");
-      }
+    // Populate a given configuration from HS2 server HiveConf, only if that configuration
```

Comment:
its not clear what the change is here. The proposed code looks similar to the existing code in the OpenSession() method. What has changed?

Reply:
The patch moves these lines (900-926) before the patch into a separate method so that we can have a retry attempt of this code. The first attempt to get the redirect URL and the second one to actually open the session.